datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1436,33 +1436,59 @@ impl<S: SimplifyInfo> TreeNodeRewriter for Simplifier<'_, S> {
 
             // CASE WHEN true THEN A ... END --> A
             // CASE WHEN X THEN A WHEN TRUE THEN B ... END --> CASE WHEN X THEN A ELSE B END
+            // CASE WHEN false THEN A END --> NULL
+            // CASE WHEN false THEN A ELSE B END --> B
+            // CASE WHEN X THEN A WHEN false THEN B END --> CASE WHEN X THEN A ELSE B END
             Expr::Case(Case {
                 expr: None,
                 mut when_then_expr,
-                else_expr: _,
+                mut else_expr,
                 // if let guard is not stabilized so we can't use it yet: https://github.com/rust-lang/rust/issues/51114
                 // Once it's supported we can avoid searching through when_then_expr twice in the below .any() and .position() calls
                 // }) if let Some(i) = when_then_expr.iter().position(|(when, _)| is_true(when.as_ref())) => {
```

Comment:
We can probably remove this comment? Not sure `if let` guards would let us refactor this given the new logic 🤔